lightning/src/ln/interactivetxs.rs

```diff
@@ -0,0 +1,1314 @@
+// This file is Copyright its original authors, visible in version control
+// history.
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+use crate::io_extras::sink;
+use crate::prelude::*;
+use core::ops::Deref;
+
+use bitcoin::blockdata::constants::WITNESS_SCALE_FACTOR;
+use bitcoin::consensus::Encodable;
+use bitcoin::policy::MAX_STANDARD_TX_WEIGHT;
+use bitcoin::{
+	absolute::LockTime as AbsoluteLockTime, OutPoint, Sequence, Transaction, TxIn, TxOut,
+};
+
+use crate::chain::chaininterface::fee_for_weight;
+use crate::events::bump_transaction::{BASE_INPUT_WEIGHT, EMPTY_SCRIPT_SIG_WEIGHT};
+use crate::ln::channel::TOTAL_BITCOIN_SUPPLY_SATOSHIS;
+use crate::ln::msgs::SerialId;
+use crate::ln::{msgs, ChannelId};
+use crate::sign::EntropySource;
+use crate::util::ser::TransactionU16LenLimited;
+
+/// The number of received `tx_add_input` messages during a negotiation at which point the
+/// negotiation MUST be failed.
+const MAX_RECEIVED_TX_ADD_INPUT_COUNT: u16 = 4096;
+
+/// The number of received `tx_add_output` messages during a negotiation at which point the
+/// negotiation MUST be failed.
+const MAX_RECEIVED_TX_ADD_OUTPUT_COUNT: u16 = 4096;
+
+/// The number of inputs or outputs that the state machine can have, before it MUST fail the
+/// negotiation.
+const MAX_INPUTS_OUTPUTS_COUNT: usize = 252;
+
+trait SerialIdExt {
+	fn is_for_initiator(&self) -> bool;
+}
+
+impl SerialIdExt for SerialId {
+	fn is_for_initiator(&self) -> bool {
+		self % 2 == 0
+	}
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum AbortReason {
+	InvalidStateTransition,
+	UnexpectedCounterpartyMessage,
+	ReceivedTooManyTxAddInputs,
+	ReceivedTooManyTxAddOutputs,
+	IncorrectInputSequenceValue,
+	IncorrectSerialIdParity,
+	SerialIdUnknown,
+	DuplicateSerialId,
+	PrevTxOutInvalid,
+	ExceededMaximumSatsAllowed,
+	ExceededNumberOfInputsOrOutputs,
+	TransactionTooLarge,
+	BelowDustLimit,
+	InvalidOutputScript,
+	InsufficientFees,
+	OutputsValueExceedsInputsValue,
+	InvalidTx,
+}
+
+#[derive(Debug)]
+pub struct TxInputWithPrevOutput {
+	input: TxIn,
+	prev_output: TxOut,
+}
+
+#[derive(Debug)]
+struct NegotiationContext {
+	holder_is_initiator: bool,
+	received_tx_add_input_count: u16,
+	received_tx_add_output_count: u16,
+	inputs: HashMap<SerialId, TxInputWithPrevOutput>,
+	prevtx_outpoints: HashSet<OutPoint>,
+	outputs: HashMap<SerialId, TxOut>,
+	tx_locktime: AbsoluteLockTime,
+	feerate_sat_per_kw: u32,
+	to_remote_value: u64,
+}
+
+impl NegotiationContext {
+	fn is_serial_id_valid_for_counterparty(&self, serial_id: &SerialId) -> bool {
+		// A received `SerialId`'s parity must match the role of the counterparty.
+		self.holder_is_initiator == !serial_id.is_for_initiator()
```

Comment:
This might be easier to read if `SerialIdExt` also had a `is_for_counterparty` method.